tools/chapel-py/scripts/generate-pyi.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python3
+#
+# Copyright 2020-2023 Hewlett Packard Enterprise Development LP
+# Copyright 2004-2019 Cray Inc.
+# Other additional copyright holders may be indicated within.
+#
+# The entirety of this work is licensed under the Apache License,
+# Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License.
+#
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import chapel.core
+from typing import List, Tuple
+
+
+def get_base_header() -> str:
+    c = chapel.core.Context()
+    return c._get_pyi_file()
+
+
+def _wrap_docstring(docstring: str, indent: int = 8) -> List[str]:
```

Comment:
As far as I can tell these are never called with different indents, so I suggest hardcoding a value (and putting it into a global variable to be shared between all of the functions here.

Reply:
Another thing you can do (instead of having each function accepting an indent argument) is to control indents declaratively with the following helper:

```Python
def indent(args: List[str], by = 8) -> List[str]:
  return [" " * indent + s for s in args]
```

Then:

```Python
        *_wrap_method("advance_to_next_revision"),
        *_wrap_docstring("Advance the context to the next revision"),
        *_def_body(),
```

Becomes:

```Python
    *indent([
        *_wrap_method("advance_to_next_revision"),
        *_wrap_docstring("Advance the context to the next revision"),
        *_def_body(),
    ])
```

From there, I suggest another modification to `indent`, which allows it to take an arglist of strings or lists of strings:

```Python
def indent(*args: str | List[str], by = 8) -> List[str]:
  temp = []
  for arg in arg:
    if !isinstance(arg, list):
      temp.append(arg)
    else:
      temp.extend(arg);
  return [" " * by + s for s in temp]
```

Then, you can write:

```Python
    *indent([
        _wrap_method("advance_to_next_revision"),
        _wrap_docstring("Advance the context to the next revision"),
        _def_body(),
    ])
```

in fact, you can even use `indent` to help get rid of these pesky stars:

```Python
s = indent(
        "class Context:",
        indent(
            _wrap_method("parse", args=[("path", "str")], rettype="List[AstNode]"),
            _wrap_docstring("Parse a top-level AST node from the given file"),
            _def_body()
        ),
        ...
        by = 0
)
```

Reply:
To make that read better, rename `indent` to `section` and `by` to `indent`. 

```
s = section(
        "class Context:",
        section(
            _wrap_method("parse", args=[("path", "str")], rettype="List[AstNode]"),
            _wrap_docstring("Parse a top-level AST node from the given file"),
            _def_body()
        ),
        ...
        indent = 0
)

Reply:
I like that last one the best, although it will actually need to be
```
s = section(
        "class Context:",
        section(
            _wrap_method("parse", args=[("path", "str")], rettype="List[AstNode]"),
            section(
                _wrap_docstring("Parse a top-level AST node from the given file"),
                _def_body(),
           ),
        ),
        ...
        indent = 0
)
```
Since `docstring` and `body` are at a higher indent level. But I will also be refactoring those in response to your above comment about the repetition.